web/src/main/java/org/openmrs/web/Listener.java

```diff
@@ -251,6 +258,7 @@ public void contextInitialized(ServletContextEvent event) {
 			log.error(MarkerFactory.getMarker("FATAL"), "Failed to obtain JDBC connection", e);
 		}
 		
+		initialized = true;
```

Comment:
I wonder if this is actually sufficient.  What about instead adding a boolean flag for "openmrsStarted", and setting this to true at the end of Listener.startOpenmrs(ServletContext).  And actually exposing a static public method "isOpenmrsStarted()" that returns the value of this.   And use this value to determine whether the application context is refreshed sufficiently to notify the listeners.  Because it seems like it isn't until this point that the application context is fully refreshed with all of the modules?

Reply:
NOTE:  Although I still think this is a reasonable feature to have added in, I recently saw that in the legacyui module, this module users a Component that implements ServletContextAware to achieve much the same behavior - basically registering a custom Session Listener through that mechanism.  So if it is better to roll back this change and just recommend that modules take this approach, that's probably fine too.  See:  https://github.com/openmrs/openmrs-module-legacyui/blob/master/omod/src/main/java/org/openmrs/web/WebComponentRegistrar.java#L50

@dkayiwa do you have opinions on how well this approach has worked?

Reply:
I have not heard of any complaints. It has the advantage of not requiring upgrading the openmrs platform version. And since it is already based on spring classes, it does not require checking whether the spring application context is initialised. But that being said, i have no objection to bringing this in.

Reply:
@dkayiwa  / @ibacher  - I've committed an update to this PR that applies the idea I laid out above.  Can you have a look and let me know what you think?  @ibacher did you test the previous fix anywhere in particular?

Reply:
I think you can just build and start the Docker image to try it out. It consistently logs a warning when starting up so if the warning is gone, it should be good.

Reply:
@ibacher is just the Docker file included with core sufficient, or do I need to use the refapp image as described on the ticket?

Reply:
It should be fine as long as we backport this to the 2.5.x branch (since the original change is also in 2.5.x).